### PR TITLE
Feat: Add streaming support for QueryEngine 

### PR DESCRIPTION
**Pull request by marcusschiesser**
- Unify streaming interface with ChatEngine (use OpenAI method)
- Add parameter object for ChatEngine and QueryEngine methods

### CODE REVIEW

File: apps/docs/docs/modules/high_level/chat_engine.md [hunk 1]
```diff
@@ -11,7 +11,16 @@ const retriever = index.asRetriever();
...
-const response = await chatEngine.chat(query);
+const response = await chatEngine.chat({ message: query });
+```
+
+The `chat` function also supports streaming, just add `stream: true` as an option:
+
+```typescript
+const stream = await chatEngine.chat({ message: query, stream: true });
+for await (const chunk of stream) {
+  process.stdout.write(chunk.response);
```
Comment (marcusschiesser) [inline apps/docs/docs/modules/high_level/chat_engine.md:22]: @yisding @himself65 I am streaming here (and in `QueryEngine`) `Response` objects and store the delta in the `response` string.

We might alternatively:
1. Use a dedicated interface (e.g. `ResponseChunk`) containing the deltas (like we're doing it for the LLM interface)
2. Extend the `Response` interface with a `delta` 

What are your thoughts?

Comment (marcusschiesser) [reply apps/docs/docs/modules/high_level/chat_engine.md:22]: Also, we might want to rename the `message` parameter. This was the original name, but now that it's a named parameter, it feels kind of weird, as it's just a `string` and not a message object.

File: packages/core/src/llm/utils.ts [hunk 2]
```diff
@@ -8,3 +6,19 @@ export async function* streamConverter<S, D>(
     yield converter(data);
   }
 }
+
+export async function* streamReducer<S, D>(params: {
```
Comment (marcusschiesser) [inline packages/core/src/llm/utils.ts:10]: @himself65 @yisding I wrote these streaming helpers - they simplify much of our streaming code. Do you know a library that we can use to deal with this concern?

Comment (himself65) [reply packages/core/src/llm/utils.ts:10]: nodejs utils might have such thing but we shouldn't use that

File: apps/docs/docs/modules/high_level/chat_engine.md (hunk 1 above)
Comment (himself65) [reply apps/docs/docs/modules/high_level/chat_engine.md:22]: > we might want to rename the `message` parameter

use `text`?

**Review by himself65 — APPROVED** [review]

File: packages/core/src/llm/utils.ts (hunk 2 above)
Comment (marcusschiesser) [reply packages/core/src/llm/utils.ts:10]: ok, then we just leave it as is for now. Please have a look, those helpers are also useful for adding new LLMs

File: apps/docs/docs/modules/high_level/chat_engine.md (hunk 1 above)
Comment (marcusschiesser) [reply apps/docs/docs/modules/high_level/chat_engine.md:22]: Actually, it's not just `text` - we also have multi-modal messages. 

see a multi-modal usage:
https://github.com/run-llama/LlamaIndexTS/blob/ab9beefa0d18ef8dc775a5d1c5d253828d16df05/packages/create-llama/templates/types/streaming/nextjs/app/api/chat/route.ts#L51-L62

Comment (marcusschiesser) [reply apps/docs/docs/modules/high_level/chat_engine.md:22]: The precise naming would be `userMessageContent`, which is too wordy, and I think just `content` is a bit confusing. Let's stick with `message` for now.
